src/main/java/org/folio/consortia/service/impl/PublicationServiceImpl.java

```diff
@@ -125,32 +129,45 @@ private List<PublicationStatusError> buildErrorListFromPublicationTenantRequestE
   }
 
   void processTenantRequests(PublicationRequest publicationRequest, PublicationStatusEntity createdPublicationEntity) {
-    List<CompletableFuture<PublicationTenantRequestEntity>> futures = new ArrayList<>();
-    Semaphore semaphore = new Semaphore(MAX_ACTIVE_THREADS);
+    List<Future<PublicationTenantRequestEntity>> futures = new ArrayList<>();
+
+    ExecutorService executor = Executors.newFixedThreadPool(MAX_ACTIVE_THREADS);
 
     for (String tenantId : publicationRequest.getTenants()) {
       try {
         PublicationTenantRequestEntity ptrEntity = buildPublicationRequestEntity(publicationRequest, createdPublicationEntity, tenantId);
-
         var savedPublicationTenantRequest = savePublicationTenantRequest(ptrEntity, folioExecutionContext);
-
-        semaphore.acquire();
-        var future = executeAsyncHttpRequest(publicationRequest, tenantId, folioExecutionContext)
-          .whenComplete((ok, err) -> semaphore.release())
-          .handle((response, t) -> updatePublicationTenantRequest(response, t, savedPublicationTenantRequest, folioExecutionContext));
+        var future = executor.submit(() -> updatePublicationTenantRequest(publicationRequest, tenantId, savedPublicationTenantRequest));
```

Comment:
Runnables need to be wrapped into getRunnableWithCurrentFolioContext/getRunnableWithFolioContext to ensure setting correct context and clearing it in the end of task execution. 
Otherwise async thread is initialized with some folioExecutionContext on creation and this cached context is used by next tasks execution, which can lead to unpredictable results, for ex there was a bug already - https://issues.folio.org/browse/MODEXPW-375

Reply:
And actually I cannot find any issues with current logic. All contexts will be diifferent only by tenantId value, and every context will be built right before making tenant specific request.

Reply:
The problem with contexts is very subtle. If context is not passed directly, thread will be initialized with folioExecutionContext value which is present in time of its initialization and its value is cached and used on subsequent calls when this thread is reused. This results in usage of the first folioExecutionContexts during other calls instead of the ones related to new request. 

We already faced and fixed such issue, there's description in this pr: https://github.com/folio-org/mod-bulk-operations/pull/61  

So if current api is not enough, I think we should ask to add support for callables as well in folio-spring-base

Reply:
What's probably different in this implementation is that executor is shutdown after call and probably it'll clean up threads somehow. But i would prefer not to rely on this underlying implementations and always pass the context